### PR TITLE
Add deep copy of handler results to avoid unwanted mutations

### PR DESCRIPTION
**Pull request by silv-io**
In instances where the result of API calls needs to split up into body and header fields, the result given by the handlers is edited in place. This can lead to unwanted mutations which affect the service state. By using a deep copy of the result for serialization this can easily be avoided.

### CODE REVIEW

**Review by thrau — CHANGES_REQUESTED** [review]
thanks for the fix silvio! i have a few comments

TL;DR: please add a test; i think we should move this to here (with a TODO to improve the implementation to not need `del`) https://github.com/localstack/localstack/blob/70d204c419cbf363fc726b468008415b45309977/localstack/aws/protocol/serializer.py#L552-L553

---

1. i followed the discussion on slack and think i understand why it's necessary, but can we add a test to demonstrate the behavior?
2. the serializer would be the better level of abstraction to do this operation, since we may be calling the serializer without the skeleton, and it's in the serialziers responsibility not to have side effects unless it's in the contract
3. performing expensive operations like making deep copies of objects for each service request should give us pause to think whether there may be different solutions. ideally we implement something where we don't need to modify anything at all, this seems to be the culprit: https://github.com/localstack/localstack/blob/70d204c419cbf363fc726b468008415b45309977/localstack/aws/protocol/serializer.py#L640 it seems we're only doing this out of convenience due to the way we organize the methods, but could probably be done without copying. maybe we can leave this as a TODO for when we notice the copy approach becomes a problem.

**Comment by silv-io** [comment]
> thanks for the fix silvio! i have a few comments
> 
> TL;DR: please add a test; i think we should move this to here (with a TODO to improve the implementation to not need `del`)
> 
> https://github.com/localstack/localstack/blob/70d204c419cbf363fc726b468008415b45309977/localstack/aws/protocol/serializer.py#L552-L553
> 
>     1. i followed the discussion on slack and think i understand why it's necessary, but can we add a test to demonstrate the behavior?
> 
>     2. the serializer would be the better level of abstraction to do this operation, since we may be calling the serializer without the skeleton, and it's in the serialziers responsibility not to have side effects unless it's in the contract
> 
>     3. performing expensive operations like making deep copies of objects for each service request should give us pause to think whether there may be different solutions. ideally we implement something where we don't need to modify anything at all, this seems to be the culprit: https://github.com/localstack/localstack/blob/70d204c419cbf363fc726b468008415b45309977/localstack/aws/protocol/serializer.py#L640
>         it seems we're only doing this out of convenience due to the way we organize the methods, but could probably be done without copying. maybe we can leave this as a TODO for when we notice the copy approach becomes a problem.

I also thought about the performance impact a deep copy could cause. This would be a good use-case for immutable types as proposed in [PEP-603](https://www.python.org/dev/peps/pep-0603/). With turning the returned dict immutable we could avoid future additions of in-place mutations. We could maybe also use [frozendict](https://pypi.org/project/frozendict/) for this.
